### PR TITLE
docs(adt): update RequestOptions documentation

### PR DESCRIPTION
**Pull request by azabbasi**

### CODE REVIEW

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs [hunk 1]
```diff
@@ -13,13 +9,13 @@ namespace Azure.DigitalTwins.Core
     public class RequestOptions
     {
         /// <summary>
-        /// A string representing a weak ETag for the entity that this request performs an operation against, as per RFC7232. The request's operation is performed
+        /// An Azure.Core.Etag object for the entity that this request performs an operation against, as per RFC7232. The request's operation is performed
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs:12]: `<see cref="" />`

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs [hunk 2]
```diff
@@ -13,13 +9,13 @@ namespace Azure.DigitalTwins.Core
...
         /// only if this ETag matches the value maintained by the server, indicating that the entity has not been modified since it was last retrieved.
         /// To force the operation to execute only if the entity exists, set the ETag to the wildcard character '*'. To force the operation to execute unconditionally, leave this value null.
         /// </summary>
         /// <remarks>
         /// If this value is not set, it defaults to null, and the ifMatch header will not be sent with the request. This means that update and delete will be unconditional and the operation will execute regardless of the existence of the resource.
         /// </remarks>
-        public string IfMatchEtag { get; set; } = null;
+        public ETag? IfMatchEtag { get; set; }
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs:19]: One of the feedback points was to remove the Etag suffix.

**Review by drwill-ms — CHANGES_REQUESTED** [review]
Can we see a test and or sample using it?

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs [hunk 3]
```diff
@@ -1,9 +1,7 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-using System;
-using System.Collections.Generic;
-using System.Text;
+using static System.Net.WebRequestMethods;
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs:4]: What is this using for?

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs [hunk 4]
```diff
@@ -13,13 +11,16 @@ namespace Azure.DigitalTwins.Core
     public class RequestOptions
     {
         /// <summary>
-        /// A string representing a weak ETag for the entity that this request performs an operation against, as per RFC7232. The request's operation is performed
+        /// An Etag object for the entity that this request performs an operation against, as per RFC7232. The request's operation is performed
         /// only if this ETag matches the value maintained by the server, indicating that the entity has not been modified since it was last retrieved.
-        /// To force the operation to execute only if the entity exists, set the ETag to the wildcard character '*'. To force the operation to execute unconditionally, leave this value null.
+        /// To force the operation to execute only if the entity exists, set the ETag to the wildcard character '*'. <code>new Etag("*")</code>. To force the operation to execute unconditionally, leave this value null.
         /// </summary>
+        /// <see href="https://docs.microsoft.com/en-us/dotnet/api/azure.etag?view=azure-dotnet">
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs:16]: See is not a standalone tag. It must be nested in remarks or summary.

We don't want summary to get too big, so really much of that should be moved to remarks. You can then put this bit referencing that URL in remarks.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs (hunk 3 above)
Comment (azabbasi) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs]: I don't know what you are talking about :D

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs (hunk 4 above)
Comment (drwill-ms) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs:16]: Also, when you get that much stuff in a remarks section, I think what we want to do is separate different thoughts with the `<para></para>` tags.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs [hunk 5]
```diff
@@ -49,7 +48,14 @@ public async Task DigitalTwins_Lifecycle()
 
                 // update twin
                 string updateTwin = TestAssetsHelper.GetRoomTwinUpdatePayload();
-                await client.UpdateDigitalTwinAsync(roomTwinId, updateTwin).ConfigureAwait(false);
+
+                // create request options to force the update using "*" ifMatch value
+                var reqeustOptions = new RequestOptions()
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs:53]: no need for parens in a ctor when using an initializer list.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs [hunk 6]
```diff
@@ -49,7 +48,14 @@ public async Task DigitalTwins_Lifecycle()
...
+                {
+                    IfMatch = new ETag("*")
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs:55]: It is good to have an example using *, but I'd feel better if our test coverage included an actual ETag value.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs [hunk 7]
```diff
@@ -13,13 +9,13 @@ namespace Azure.DigitalTwins.Core
     public class RequestOptions
     {
         /// <summary>
-        /// A string representing a weak ETag for the entity that this request performs an operation against, as per RFC7232. The request's operation is performed
+        /// An Etag object <see href="https://docs.microsoft.com/en-us/dotnet/api/azure.etag?view=azure-dotnet"> See the documentation for Etag. </see> for the entity that this request performs an operation against, as per RFC7232. The request's operation is performed
         /// only if this ETag matches the value maintained by the server, indicating that the entity has not been modified since it was last retrieved.
-        /// To force the operation to execute only if the entity exists, set the ETag to the wildcard character '*'. To force the operation to execute unconditionally, leave this value null.
+        /// To force the operation to execute only if the entity exists, set the ETag to the wildcard character '*'. <code>new Etag("*")</code>. To force the operation to execute unconditionally, leave this value null.
```
Comment (bikamani) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs:14]: :thumbsup:

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs [hunk 8]
```diff
@@ -49,7 +48,14 @@ public async Task DigitalTwins_Lifecycle()
...
                 string updateTwin = TestAssetsHelper.GetRoomTwinUpdatePayload();
-                await client.UpdateDigitalTwinAsync(roomTwinId, updateTwin).ConfigureAwait(false);
+
+                // create request options to force the update using "*" ifMatch value
+                var reqeustOptions = new RequestOptions
+                {
+                    IfMatch = new ETag("*")
+                };
+
+                await client.UpdateDigitalTwinAsync(roomTwinId, updateTwin, reqeustOptions).ConfigureAwait(false);
```
Comment (bikamani) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs:58]: how are we verifying?

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs [hunk 9]
```diff
@@ -13,13 +9,18 @@ namespace Azure.DigitalTwins.Core
...
+        /// The request's operation is performed only if this ETag matches the value maintained by the server,
+        /// indicating that the entity has not been modified since it was last retrieved.
+        /// To force the operation to execute only if the entity exists, set the ETag to the wildcard character '*'.
+        /// To force the operation to execute unconditionally, leave this value null.
+        /// If this value is not set, it defaults to null, and the ifMatch header will not be sent with the request.
+        /// This means that update and delete will be unconditional and the operation will execute regardless of the existence of the resource.
+        /// </para>
         /// </remarks>
-        public string IfMatchEtag { get; set; } = null;
+        public string IfMatch { get; set; } = null;
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/RequestOptions.cs:24]: I know this was already set this way, but string already defaults to null, so no reason to initialize.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs [hunk 10]
```diff
@@ -49,7 +48,14 @@ public async Task DigitalTwins_Lifecycle()
...
                 string updateTwin = TestAssetsHelper.GetRoomTwinUpdatePayload();
-                await client.UpdateDigitalTwinAsync(roomTwinId, updateTwin).ConfigureAwait(false);
+
+                // create request options to force the update using "*" ifMatch value
+                var reqeustOptions = new RequestOptions
+                {
+                    IfMatch = "*"
+                };
+
+                await client.UpdateDigitalTwinAsync(roomTwinId, updateTwin, reqeustOptions).ConfigureAwait(false);
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs:58]: requestOptions is misspelled

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs (hunk 8 above)
Comment (drwill-ms) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs:58]: The request would be rejected and throw, if it failed.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs [hunk 11]
```diff
@@ -49,7 +48,14 @@ public async Task DigitalTwins_Lifecycle()
...
 
                 // delete a twin
                 await client.DeleteDigitalTwinAsync(roomTwinId).ConfigureAwait(false);
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs:61]: can we add a test for an actual etag value too, perhaps here with delete?

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs (hunk 8 above)
Comment (drwill-ms) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/DigitalTwinTests.cs:58]: Note we're not trying to test the service. We're just making sure we pass along a payload and it gets sent/does not explode. We could maybe do more testing of specific scenarios (e.g. create, get etag, update, delete with original etag, ensure it fails) but that ventures into testing the service.

**Review by drwill-ms — APPROVED** [review]
Fix misspelling, remove init to null, and I'm good.

Consider adding the actual etag for delete, or we can do it later.

**Comment by drwill-ms** [comment]
I updated your PR title. You'll want to make sure the merge commit message is updated as well.

**Pull request merged**